src/descriptor/tr.rs

```diff
@@ -0,0 +1,279 @@
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let desc = self.to_string_no_checksum();
+        write!(f, "{}", &desc)
+    }
+}
+
+impl<Pk: MiniscriptKey> Tr<Pk> {
+    fn taptree_height(tree: &TapTree<Pk>) -> usize {
+        match tree {
+            TapTree::Tree(ref left_tree, ref right_tree) => {
+                1 + max(
+                    Self::taptree_height(&**left_tree),
+                    Self::taptree_height(&**right_tree),
+                )
+            }
+            TapTree::Leaf(_) => 1,
+        }
+    }
+
+    pub fn new(internal_key: Pk, tree: Option<TapTree<Pk>>) -> Result<Self, Error> {
+        let nodes = match tree {
+            Some(ref t) => Self::taptree_height(t),
+            None => 0,
+        };
+
+        if nodes > TAPROOT_MAX_NODE_COUNT {
+            Ok(Self { internal_key, tree })
+        } else {
+            Err(Error::MaxRecursiveDepthExceeded)
+        }
+    }
+
+    fn to_string_no_checksum(&self) -> String {
+        let key = &self.internal_key;
+        match self.tree {
+            Some(ref s) => format!("tr({},{})", key, s),
+            None => format!("tr({})", key),
+        }
+    }
+
+    pub fn get_internal_key(&self) -> Result<Pk, Error> {
+        Ok(self.internal_key.clone())
+    }
+
+    pub fn get_taptree(&self) -> Result<TapTree<Pk>, Error> {
+        match &self.tree {
+            Some(t) => Ok(t.clone()),
+            None => Err(Error::Unexpected("TapTree empty".to_string())),
+        }
+    }
+}
+
+impl<Pk> Tr<Pk>
+where
+    Pk: MiniscriptKey + FromStr,
+    Pk::Hash: FromStr,
+    <Pk as FromStr>::Err: ToString,
+    <<Pk as MiniscriptKey>::Hash as FromStr>::Err: ToString,
+{
+    fn parse_miniscript(script: &str) -> Result<Miniscript<Pk, Segwitv0>, Error> {
+        let (ref script_tree, rest) = Tree::from_slice(script)?;
+        if rest.is_empty() {
+            Miniscript::<Pk, Segwitv0>::from_tree(script_tree)
+        } else {
+            return Err(Error::Unexpected(format!(
+                "error parsing miniscript from tapscript"
+            )));
+        }
+    }
+
+    // helper function for semantic parsing of script paths
+    pub fn tr_script_path(tree: &Tree) -> Result<TapTree<Pk>, Error> {
+        match tree {
+            Tree { name, args } if name.len() > 0 && args.len() == 0 => {
+                let script = name;
+                let script = Self::parse_miniscript(script)?;
+                let script = Arc::new(script);
+                Ok(TapTree::Leaf(script))
+            }
+            Tree { name, args } if name.len() == 0 && args.len() == 2 => {
+                let left_branch = &args[0];
+                let right_branch = &args[1];
+                let left_tree = Self::tr_script_path(&left_branch)?;
+                let right_tree = Self::tr_script_path(&right_branch)?;
+                let left_ref = Arc::new(left_tree);
+                let right_ref = Arc::new(right_tree);
```

Comment:
nit: Prefer inlining some code instead of 6 lines. 

```
let left = Self::tr_script_path(&args[0])?;
..
Ok(TapTree::Tree(Arc::new(left_ref), Arc::new(right_ref)))
```

Reply:
Will do! Besides this, refactoring this function to a method should be more idiomatic, right?